test/ApplicationRecordLogTest.cpp

```diff
@@ -0,0 +1,383 @@
+    m_record_log->exit(hash, time_exit1);
+    m_record_log->enter(hash, time_entry2);
+    m_record_log->exit(hash, time_exit2);
+    m_record_log->dump(records, short_regions);
+
+    ASSERT_EQ(1ULL, records.size());
+    EXPECT_EQ(1.0, records[0].time);
+    EXPECT_EQ(proc_id, records[0].process);
+    EXPECT_EQ(geopm::EVENT_SHORT_REGION, records[0].event);
+    EXPECT_EQ(0ULL, records[0].signal);
+    ASSERT_EQ(1ULL, short_regions.size());
+    EXPECT_EQ(hash, short_regions[0].hash);
+    EXPECT_EQ(2, short_regions[0].num_complete);
+    EXPECT_EQ(3.0, short_regions[0].total_time);
+}
+
+TEST_F(ApplicationRecordLogTest, dump_twice)
+{
+    std::vector<record_s> records;
+    std::vector<short_region_s> short_regions;
+    int proc_id = 123;
+    // Note time_zero is one second after 1970
+    geopm_time_s time_0 = {{1, 0}};
+    geopm_time_s time_1 = {{2, 0}};
+    geopm_time_s time_2 = {{3, 0}};
+
+    m_record_log->set_process(proc_id);
+    m_record_log->set_time_zero(time_0);
+
+    m_record_log->enter(0x1234, time_1);
+    m_record_log->exit(0x1234, time_2);
+    m_record_log->epoch(time_2);
+
+    m_record_log->dump(records, short_regions);
+    EXPECT_EQ(2ULL, records.size());
+    EXPECT_EQ(1ULL, short_regions.size());
+
+    m_record_log->dump(records, short_regions);
+    EXPECT_EQ(0ULL, records.size());
+    EXPECT_EQ(0ULL, short_regions.size());
+}
+
+TEST_F(ApplicationRecordLogTest, dump_within_region)
+{
+    std::vector<record_s> records;
+    std::vector<short_region_s> short_regions;
+    int proc_id = 123;
+    uint64_t hash = 0xABCD;
+    // Note time_zero is one second after 1970
+    geopm_time_s time_0 = {{1, 0}};
+    m_record_log->set_process(proc_id);
+    m_record_log->set_time_zero(time_0);
+
+    m_record_log->enter(hash, {2, 0});
+    m_record_log->exit(hash, {3, 0});
+    m_record_log->enter(hash, {4, 0});
+    m_record_log->epoch({5, 0});
+    m_record_log->dump(records, short_regions);
+    ASSERT_EQ(2ULL, records.size());
+    EXPECT_EQ(geopm::EVENT_SHORT_REGION, records[0].event);
+    EXPECT_EQ(geopm::EVENT_EPOCH_COUNT, records[1].event);
+    EXPECT_EQ(0ULL, records[0].signal);  // short region index
+    EXPECT_EQ(1ULL, records[1].signal);  // epoch count
+    ASSERT_EQ(1ULL, short_regions.size());
+    EXPECT_EQ(hash, short_regions[0].hash);
+    EXPECT_EQ(1, short_regions[0].num_complete);
+    EXPECT_EQ(1.0, short_regions[0].total_time);
+
+    m_record_log->epoch({6, 0});
+    m_record_log->dump(records, short_regions);
+    ASSERT_EQ(1ULL, records.size());
+    EXPECT_EQ(geopm::EVENT_EPOCH_COUNT, records[0].event);
+    EXPECT_EQ(2ULL, records[0].signal);
+    ASSERT_EQ(0ULL, short_regions.size());
+
+    m_record_log->epoch({7, 0});
+    m_record_log->exit(hash, {8, 0});
+    m_record_log->dump(records, short_regions);
+    ASSERT_EQ(2ULL, records.size());
+    EXPECT_EQ(geopm::EVENT_EPOCH_COUNT, records[0].event);
+    EXPECT_EQ(geopm::EVENT_SHORT_REGION, records[1].event);
```

Comment:
Your explanation last week helped me understand this. Maybe a similar comment here might help draw attention to the significance of this line. If I understand correctly, this is expected to be promoted to a short region because the finishing region in this dump was started as any instance after a full short-running-region cycle for that hash while still in the previous dump.